More tempo line fixes

mike-spa opened the pull request:
Resolves: #14296

Fixing the next bunch of issues with tempo lines.
- Fixed wrong staff on copy-pasting
- Fixed mid-measure tempo indication disappearing in parts
- ~Fixed wrong staff of tempo lines upon reordering instruments~
- ~Fixed lost tempo lines upon save-reopen if they start or end mid-measure~

The layout of tempo lines (actually, all system-level text lines) is still imperfect because (let's all repeat together):

_We cAN't AtTAcH THinGs To RhyTHmIC PoSiTiONS_

Most commonly: things will just snap to the start/end of a measure if there is nothing to attach to within that measure. Can't really do anything about that at the moment.

cbjeukendrup reviewed (commented):
The code looks good to me, although I must admit that I don't know this part of the code well enough to understand all the details behind every change.

mike-spa commented:
UPDATE
I have removed a couple of commits from this PR.
1. The PR does _not_ try anymore to solve the last issue mentioned [here](https://github.com/musescore/MuseScore/pull/15931#issue-1536188563) (see Casper's comment [here](https://github.com/musescore/MuseScore/issues/14296#issuecomment-1381839350) for a more detailed description of the issue itself). The root cause of the issue is that, in this scenario, the line ends up with no valid start- or end-segment, so it will also not be written in the file. My previous solution tried to fix the problem by changing how the writing of spanners works, but a) It's too dangerous, b) Because the writing of spanners changed, it was causing dozens and dozens of utest failures, c) It is really not the solution we ultimately want. What we want is proper attachment to rythmic position, so that the line does _not_ end up with invalid start- or end anchors. And, as we know, this is part of a larger job. So we'll just have to live with that issue until we can fix it properly.
2. The second-to-last issue mentioned [here](https://github.com/musescore/MuseScore/pull/15931#issue-1536188563) was meanwhile solved by [another more complete PR](https://github.com/musescore/MuseScore/pull/15938) so it wasn't necessary anymore. 

@oktophonie please do take a look if you can, and check that everything else works as intended. @RomanPudashkin the tests are now fixed so I think it's ready for review

RomanPudashkin commented on src/engraving/layout/layoutmeasure.cpp:
```diff
@@ -501,7 +515,28 @@ static bool breakMultiMeasureRest(const LayoutContext& ctx, Measure* m)
+                }
+                if ((e->isRehearsalMark()
+                     || e->isTempoText()
+                     || ((e->isHarmony() || e->isStaffText() || e->isSystemText() || e->isTripletFeel() || e->isPlayTechAnnotation()
```
This block of code is impossible to read/debug. I think you could use static std::set/unordered_set here instead of adding a lot of checks to this condition

RomanPudashkin commented on src/engraving/layout/layoutmeasure.cpp:
```diff
@@ -459,14 +459,28 @@ static bool breakMultiMeasureRest(const LayoutContext& ctx, Measure* m)
+        for (auto i : prevMeasSpanners) {
+            Spanner* s = i.value;
+            if ((s->isVolta() || s->isGradualTempoChange() || s->isTextLine())
+                && ((s->tick2() > prevMeas->tick() && s->tick2() < prevMeas->endTick())
```
Same as https://github.com/musescore/MuseScore/pull/15931/files#r1115733878

+ a lot of redundant calls (for example, prevMeas->tick() is called twice here). This also creates performance problems

RomanPudashkin commented on src/engraving/libmscore/excerpt.cpp:
```diff
@@ -1420,6 +1420,21 @@ void Excerpt::cloneStaff2(Staff* srcStaff, Staff* dstStaff, const Fraction& star
+                for (EngravingItem* e : oseg->annotations()) {
+                    if (e->generated()
+                        || (e->track() != srcTrack && !(e->systemFlag() && e->track() == 0)) // system items must be cloned even if they are on different tracks
+                        || (e->track() != srcTrack && e->systemFlag() && e->findLinkedInScore(score))) { // ...but only once!
```
This condition has become even more complicated than it was before
